thetis/timeintegrator.py

```diff
@@ -804,6 +1277,27 @@ def __init__(self, equation, dt, solver_parameters={}, terms_to_add='all'):
                                             solver_parameters, terms_to_add)
 
 
+class BackwardEulerNew(DIRKGenericNew):
+    """
+    Backward Euler method
+
+    This method has the Butcher tableau
+
+    1   | 1
+    ---------
+        | 1
+    """
+    def __init__(self, equation, solution, fields, dt,
+                 bnd_conditions=None, solver_parameters={}):
+        a = [[1.0]]
+        b = [1.0]
+        c = [1.0]
+        super(BackwardEulerNew, self).__init__(equation, solution, fields, dt,
+                                               a, b, c,
+                                               bnd_conditions=bnd_conditions,
+                                               solver_parameters=solver_parameters)
+
+
```

Comment:
The only thing that these dirk classes do is define the tableau.  You can simplify the code like this:

```python

class DIRKGeneric(object):
    
    def __init__(self, equation, solution, fields, dt, bnd_conditions=None, solver_parameters={})
        self.n_stages = len(self.b)
        ...


class BackwardEuler(DIRKGeneric):
    a = [[1.0]]
    b = [1.0]
    c = [1.0]

class DIRKLSPUM2(DIRKGeneric):
     a = [[2.0/11.0, 0, 0],
          [205.0/462.0, 2.0/11.0, 0],
          [2033.0/4620.0, 21.0/110.0, 2.0/11.0]]
     b = [24.0/55.0, 1.0/5.0, 4.0/11.0]
     c = [2.0/11.0, 289.0/462.0, 751.0/924.0]
```

IOW, you can set the Butcher coefficients as class attributes in the concrete classes and just use them directly generic class.  If you wanted to be really fancy you could require that the subclasses set those coefficients like so:

```python
from abc import ABCMeta, abstractproperty

class DIRK(object):
    __metaclass__ = ABCMeta

    @abstractproperty
    def a(self):
        pass

    @abstractproperty
    def b(self):
        pass

    @abstractproperty
    def c(self):
        pass

class MyDIRK(DIRK):
    a = ...
    b = ...
    c = ...
```

Now any class derived from `DIRK` cannot be instantiated unless it implements all of `a`, `b` and `c`.